packages/resolver/src/adapters/cns-utils.ts

```diff
@@ -0,0 +1,146 @@
+}
+
+export const validateVirtualSubdomainEnabled = (
+  metadata: Api.Cardano.NftMetadata,
+) => {
+  const {virtualSubdomainEnabled} = metadata
+  return virtualSubdomainEnabled === 'Enabled'
+}
+
+/* istanbul ignore next */
+export const stringToHex = (str: string) =>
+  Buffer.from(str, 'utf8').toString('hex')
+/* istanbul ignore next */
+export const hexToString = (hex: string) =>
+  Buffer.from(hex, 'hex').toString('utf8')
+
+export const parseAssocMapAsync = async <T>(
+  assocMapVal: AssocMap<BuiltinByteString, T>,
+  itemParser: (args0: T) => Promise<string>,
+  limit = 5,
+): Promise<string[][]> => {
+  const parsedAssocMap: string[][] = []
+  const promises: Promise<string>[] = []
+
+  for (let i = 0; i < limit; i += 1) {
+    if (i >= assocMapVal.map.length) continue
+    const mapItem = assocMapVal.map[i]
+    if (!mapItem) throw new Error('bad data')
+    promises.push(itemParser(mapItem.v))
+  }
+
+  const valueArray = await Promise.all(promises)
+  for (let i = 0; i < limit; i += 1) {
+    if (i >= assocMapVal.map.length) continue
+    const mapItem = assocMapVal.map[i]
+    /* istanbul ignore next */
+    if (!mapItem) throw new Error('bad data')
+    const key = hexToString(mapItem.k.bytes)
+    const value = valueArray[i]
+    if (!value) throw new Error('bad data')
+    parsedAssocMap.push([key, value])
+  }
+  return parsedAssocMap
+}
+
+export const objToHex = async <T>(
+  obj: T,
+  csl: WasmModuleProxy,
+): Promise<string> => {
+  const plutusData = await csl.PlutusData.fromJson(JSON.stringify(obj), 1)
+  const result = await plutusData.toHex()
+  return result
+}
+
+export const parsePlutusAddressToBech32 = async (
+  plutusHex: string,
+  csl: WasmModuleProxy,
+  networkId: number,
+): Promise<string> => {
+  const cslPlutusDataAddress = await csl.PlutusData.fromHex(plutusHex)
+  const plutusDataAddressJson = await cslPlutusDataAddress.toJson(1)
+  const plutusDataAddressObject = JSON.parse(plutusDataAddressJson)
+  const plutusDataPaymentKeyObject = plutusDataAddressObject.fields[0]
+  const plutusDataStakeKeyObject = plutusDataAddressObject.fields[1]
+  const cslPaymentKeyHash = plutusDataPaymentKeyObject.fields[0].bytes
+
+  // Take into account whether the hash is a PubKeyHash or ScriptHash
+  const credentialKeyHashBytes = Buffer.from(cslPaymentKeyHash, 'hex')
+  const pubKeyCredentialKeyHash = await csl.Ed25519KeyHash.fromBytes(
+    credentialKeyHashBytes,
+  )
+  const pubKeyCredential = await csl.Credential.fromKeyhash(
+    pubKeyCredentialKeyHash,
+  )
+  const scriptCredentialKeyHash = await csl.ScriptHash.fromBytes(
+    credentialKeyHashBytes,
+  )
+  const scriptCredential = await csl.Credential.fromScripthash(
+    scriptCredentialKeyHash,
+  )
+  const cslPaymentCredential =
+    plutusDataPaymentKeyObject.constructor === 0
+      ? pubKeyCredential
+      : scriptCredential
+  let bech32Addr = ''
+
+  // Parsing address according to whether it has a stake key
+  /* istanbul ignore else */
```

Comment:
[this](https://github.com/Emurgo/yoroi/pull/2959/files#diff-747c1031314b9fe4f77827ffe371122855f8eb331d36f8bcb6e4b89377dbcb7cR119) line is in contradiction with this [line](https://github.com/Emurgo/yoroi/pull/2959/files#diff-747c1031314b9fe4f77827ffe371122855f8eb331d36f8bcb6e4b89377dbcb7cR93)

need to ask hinson

Reply:
ask him to review this part.